RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -398,43 +789,123 @@ namespace {
+      if (hitsInLayer.begin()->estChi2 < btlChi2Cut_ && hitsInLayer.begin()->timeChi2 < btlTimeChi2Cut_) {
+        hitMatched = true;
+        output.push_back(hitbuilder->build(hitsInLayer.begin()->hit));
+        if (*(hitsInLayer.begin()) < bestHit)
+          bestHit = *(hitsInLayer.begin());
+      }
+    }
+
+    if (useVertex_ && matchVertex && !hitMatched) {
+      //try a second search with beamspot hypothesis
+      hitsInLayer.clear();
+      find_hits_in_dets(hits, traj, ilay, tsos, 0., bs, prop, *theEstimator, *hitbuilder, false, hitsInLayer);
+      if (!hitsInLayer.empty()) {
+        if (hitsInLayer.begin()->timeChi2 < btlTimeChi2Cut_) {
+          if (hitsInLayer.begin()->estChi2 < btlChi2Cut_) {
+            hitMatched = true;
+            output.push_back(hitbuilder->build(hitsInLayer.begin()->hit));
+            if ((*hitsInLayer.begin()) < bestHit)
+              bestHit = *(hitsInLayer.begin());
+          }
+        }
+      }
+    }
   }
   return output;
 }
 
 template <class TrackCollection>
 TransientTrackingRecHit::ConstRecHitContainer TrackExtenderWithMTDT<TrackCollection>::tryETLLayers(
     const TrackType& track,
+    const Trajectory& traj,
     const MTDTrackingDetSetVector& hits,
     const MTDDetLayerGeometry* geo,
     const MagneticField* field,
-    const Propagator* prop) const {
+    const Propagator* prop,
+    const reco::BeamSpot& bs,
+    const double& vtxTime,
+    const bool matchVertex,
+    MTDHitMatchingInfo& bestHit) const {
   TransientTrackingRecHit::ConstRecHitContainer output;
   const vector<const DetLayer*>& layers = geo->allETLLayers();
 
+  bestHit = MTDHitMatchingInfo();
   auto tTrack = builder->build(track);
+  // get the outermost trajectory point on the track
+  TrajectoryStateOnSurface tsos = tTrack.outermostMeasurementState();
 
   for (const DetLayer* ilay : layers) {
     const BoundDisk& disk = static_cast<const MTDRingForwardDoubleLayer*>(ilay)->specificSurface();
     const double diskZ = disk.position().z();
 
-    // get the outermost trajectory point on the track
-    TrajectoryStateOnSurface tsos = tTrack.outermostMeasurementState();
     if (tsos.globalPosition().z() * diskZ < 0)
       continue;  // only propagate to the disk that's on the same side
-    find_hits_in_dets(hits, ilay, tsos, prop, *theEstimator, *hitbuilder, output);
+
+    std::set<MTDHitMatchingInfo> hitsInLayer;
+    bool hitMatched = false;
+
+    if (useVertex_ && matchVertex)
+      find_hits_in_dets(hits, traj, ilay, tsos, vtxTime, bs, prop, *theEstimator, *hitbuilder, true, hitsInLayer);
+    else
+      find_hits_in_dets(hits, traj, ilay, tsos, 0., bs, prop, *theEstimator, *hitbuilder, false, hitsInLayer);
+
+    //just take the first hit because the hits are sorted on their matching quality
+    if (!hitsInLayer.empty()) {
+      //check hits to pass minimum quality matching requirements
+      if (hitsInLayer.begin()->estChi2 < btlChi2Cut_ && hitsInLayer.begin()->timeChi2 < btlTimeChi2Cut_) {
```

Comment:
this is in tryETLLayers, why is btlChi2Cut_ used?
It looks like the implementation is identical to the BTL part. Please refactor to eliminate code replication